main.go

```diff
@@ -0,0 +1,377 @@
+		})
+
+		if curCu > maxCu {
+			maxCu = curCu
+		}
+
+		if prevRps > report.Rps {
+			printers.doPrint(func(printer Printer) {
+				printer.PrintEvent("RPS DECREASED")
+			})
+
+			if options.StopOnEvents {
+				break
+			}
+		}
+
+		if prevMean*100 < report.Average {
+			printers.doPrint(func(printer Printer) {
+				printer.PrintEvent("LATENCY INCREASED")
+			})
+
+			if options.StopOnEvents {
+				break
+			}
+		}
+
+		if succRate < 0.85 {
+			printers.doPrint(func(printer Printer) {
+				printer.PrintEvent("ERROR RATE INCREASED")
+			})
+
+			if options.StopOnEvents {
+				break
+			}
+		}
+
+		prevRps = report.Rps
+		prevMean = report.Average
+	}
+
+	maxCu = maxCu / options.StepDuration
+
+	printers.doPrint(func(printer Printer) {
+		printer.PrintFooter(maxCu)
+	})
+}
+
+func calcErrorRate(dist map[string]int) float64 {
+	total := 0.0
+	nok := 0.0
+	for k, v := range dist {
+		if k == "Canceled" {
+			continue
+		}
+		total += float64(v)
+		if k != "OK" {
+			nok += float64(v)
+		}
+	}
+	return 1 - (nok / total)
+}
+
+func NewEthSpamBinaryDataFunc(queryParams map[string]int64, parentHost string, chain dshackle.ChainRef, cuCount *uint64, ctx context.Context) func(mtd *desc.MethodDescriptor, callData *runner.CallData) []byte {
+	generator, err := ethspam.MakeQueriesGenerator(queryParams)
+	if err != nil {
+		exit(1, "failed to install defaults: %s", err)
+	}
+
+	client, err := node.NewClient(ctx, parentHost)
+	if err != nil {
+		exit(1, "failed to make a new client: %s", err)
+	}
+	mkState := ethspam.StateProducer{
+		Client: client,
+	}
+
+	stateChannel := make(chan ethspam.State, 1)
+
+	randSrc := rand.NewSource(time.Now().UnixNano())
+	go func() {
+		state := ethspam.LiveState{
+			IdGen:   &ethspam.IdGenerator{},
+			RandSrc: randSrc,
+		}
+		for {
+			newState, err := mkState.Refresh(&state)
+			if err != nil {
+				// It can happen in some testnets that most of the blocks
+				// are empty(no transaction included), don't refresh the
+				// QueriesGenerator state without new inclusion.
+				if err == ethspam.ErrEmptyBlock {
+					select {
+					case <-time.After(5 * time.Second):
+					case <-ctx.Done():
+						return
+					}
+					continue
+				}
+				fmt.Printf("failed to refresh state: %s", err)
+				<-time.After(1 * time.Second)
+				continue
+			}
+			select {
+			case stateChannel <- newState:
+			case <-ctx.Done():
+				return
+			}
+
+			select {
+			case <-time.After(15 * time.Second):
+			case <-ctx.Done():
+			}
+		}
+	}()
+
+	state := <-stateChannel
+
+	queries := make(chan ethspam.QueryContent, 1000)
+
+	go func() {
+		for {
+			// Update state when a new one is emitted
+			select {
+			case state = <-stateChannel:
+			case <-ctx.Done():
+				return
+			default:
+			}
+			if q, err := generator.Query(state); err == io.EOF {
+				return
+			} else if err != nil {
+				exit(2, "failed to write generated query: %s", err)
+			} else {
+				queries <- q
```

Comment:
We can hang here, and this will lead goroutines leak. The best solution is to add another select here.